samza-core/src/main/java/org/apache/samza/zk/ZkUtils.java

```diff
@@ -65,13 +65,7 @@
   private volatile String ephemeralPath = null;
   private final ZkKeyBuilder keyBuilder;
   private final int connectionTimeoutMs;
-  private ZkJobCoordinatorMetrics metrics;
-
-  public ZkUtils(ZkKeyBuilder zkKeyBuilder, ZkClient zkClient, int connectionTimeoutMs) {
-    this.keyBuilder = zkKeyBuilder;
-    this.connectionTimeoutMs = connectionTimeoutMs;
-    this.zkClient = zkClient;
-  }
+  private final ZkJobCoordinatorMetrics metrics;
 
   public ZkUtils(ZkKeyBuilder zkKeyBuilder, ZkClient zkClient, int connectionTimeoutMs, ZkJobCoordinatorMetrics metrics) {
```

Comment:
> Just to fix a NPE, I didn’t want to get into refactoring stuff.

Yes. fixing NPE is what this PR is about. If you are thinking of refactoring how metrics are wired in the new APIs, let's solve it in a separate PR. 

> Though we achieve isolation in terms of registering/creating metics and publishing metrics , following this pattern introduces dependency between different layers.
> Basic premise here is each layer expects the previous layer creating it to have a metricsRegistry(map) to pass it to current layer. There’s a fan out from top layer (who registers with reporters) to all lower level.

Yes. That's right. If the dependencies are skipping across layers, then it is perhaps an opportunity to step-back and revisit the component layout itself, rather than directly solving for dependency injection.

> In chain A-B-C if A registers metricsRegistry with reporters and B doesn’t need MetricsRegistry for reporting, it still has to have it as explicit dependency
since C might require it to report metrics(which it should receive from B).

Yes. That's right. If A manages the component B and B manages C, B will be a passthrough for any dependency of component C on A. There isn't another way around it, afaik unless we have a separate dependency injection system.